### PR TITLE
WIP: pyjulia updates

### PR DESCRIPTION
**Pull request by Keno**
This does two things:

1. A rebase of some of the previous pull requests to bring pyjulia up to date with more modern versions of julia
2. Attempt to solve the challenges encountered with PyCall (i.e. #51)

The approach taken to 2) is to

a. Separate out cache files into a separate directory to reflect the fact that this is a quite different environment. It may be theoretically possible to share packages that do not depend on PyCall, but this could turn out to be quite brittle and the necessary logic does not currently exist in julia to handle this situation. I think a clean separation as is done here is the best way to go.
b. Use a python script that calls the argument parsing routines directly. This is necessary because the precompilation and runtime environment needs to be the same in order for PyCall to be able to be used without major modification (small PyCall PR to enable loading in-process symbols is forthcoming). I played with a couple of variants here, but I think I've arrived at a very simple/effective solution.

Remaining todo:
- [x] Only enable this if python is statically linked?
- [x] COFF Wrapper for windows
- [x] Test on other Unix platforms
- [x] Mac CI
- [x] Enable AppVeyor
- [x] Make sure new code path is actually tested
- [x] Windows cross-version test on CI
- [x] Update COFF wrappers to 32bit
- [x] https://github.com/JuliaLang/julia/pull/17458 to fix tests

### CODE REVIEW

File: julia/core.py [hunk 1]
```diff
@@ -297,6 +297,14 @@ def __init__(self, init_julia=True, jl_runtime_path=None, jl_init_path=None,
         self.api.jl_exception_clear()
 
         if init_julia:
+            # Replace the cache directory with a private one. PyCall needs a different
+            # configuration and so do any packages that depend on it. Ideally, we could
```
Comment (tkelman) [inline julia/core.py]: why can't we use the normal pycall?

Comment (Keno) [inline julia/core.py]: Needs a different configuration (in-process vs shared library).

File: julia/core.py [hunk 2]
```diff
@@ -271,42 +280,66 @@ def __init__(self, init_julia=True, jl_init_path=None):
...
+            os.environ["PYCALL_PYTHON_EXE"] = sys.executable
+            os.environ["PYCALL_JULIA_HOME"] = JULIA_HOME
+            self._call(u"eval(Base,:(JULIA_HOME=\"/Users/kfischer/Projects/pyjulia/fake-julia\"))")
+            # Replace the cache directory with a private one. PyCall needs a different
+            # configuration and so do any packages that depend on it. Ideally, we could
+            # detect packages that depend on PyCall and only use LOAD_CACHE_PATH for them
+            # but that would be significantly more complicated and brittle, and may not
+            # be worth it.
+            self._call(u"empty!(Base.LOAD_CACHE_PATH)")
+            self._call(u"push!(Base.LOAD_CACHE_PATH, abspath(Pkg.Dir._pkgroot()," +
```
Comment (tkelman) [inline julia/core.py]: using internal methods here doesn't look like a great idea for stability's sake

Comment (Keno) [inline julia/core.py]: You mean `_pkgroot`? Unfortunately that's how to determine where this directory is. If you want to advocate for exporting that method from Base, be my guest.

Comment (tkelman) [inline julia/core.py]: Or use the existing `LOAD_CACHE_PATH` and just modify that.

Comment (Keno) [inline julia/core.py]: Seems like an ok option/

File: julia/core.py [hunk 3]
```diff
@@ -193,6 +193,15 @@ def module_functions(julia, module):
             pass
     return bases
 
+def determine_if_statically_linked():
+    """Determines if this python executable is statically linked"""
+    # Windows and OS X are generally always dynamically linked
+    if not sys.platform.startswith('linux'):
+        return False
+    lddoutput = subprocess.check_output(["ldd",sys.executable])
+    return not lddoutput.conains("libpython")
```
Comment (tkelman) [inline julia/core.py]: typo?

Comment (Keno) [inline julia/core.py]: Yes already fixed locally. Give me a few more mins and I'll push all the windows code.

**Comment by Keno** [comment]
There's a travis failure here I don't understand: `Type str doesn't support the buffer API`. Works fine for me locally, even with python 3.

File: appveyor.yml [hunk 4]
```diff
@@ -28,10 +28,6 @@ environment:
   - JULIAVERSION: "julialang/bin/winnt/x64/0.4/julia-0.4-latest-win64.exe"
     PYTHONDIR: "C:\\Python35-x64"
 
-branches:
-  only:
-    - clean-julia-0.4
```
Comment (tkelman) [inline appveyor.yml]: should probably change this to only master, otherwise you get duplicate builds (branch and pr)

File: .travis.yml [hunk 5]
```diff
@@ -1,14 +1,24 @@
...
 notifications:
   email: false
 before_install:
   - sudo add-apt-repository ppa:staticfloat/julia-deps -y
-  - sudo add-apt-repository ppa:staticfloat/juliareleases -y
+  if [ "$JULIA_VERSION" = "LAST_RELEASE" ]; then
+      - sudo add-apt-repository ppa:staticfloat/juliareleases -y
+  else
+      - sudo add-apt-repository ppa:staticfloat/juliareleases -y
+  end
```
Comment (tkelman) [inline .travis.yml]: We shouldn't be using the ppa here, it's effectively deprecated. Also doesn't help on osx.

Comment (Keno) [inline .travis.yml]: What's the suggested way?

Comment (tkelman) [inline .travis.yml]: could modernize https://github.com/JuliaLang/julialang.github.com/blob/master/install-julia.sh (and take out the ppa parts) which approximates what `language: julia` does on travis. Travis doesn't support multiple language settings afaik, and `language: python` is probably more complicated.

**Comment by stevengj** [comment]
Do we really want `julia-debug.exe` to be checked into git?

**Comment by Keno** [comment]
Yes, it's not the real julia-debug.exe, it's just a very small COFF wrapper.

**Comment by tkelman** [comment]
win32 or win64?

**Comment by Keno** [comment]
win64, though as long as your os is 64bit it doesn't matter. I suppose I could just compile it as 32bit and have it work anywhere.

**Comment by tkelman** [comment]
Your Travis syntax is broken here and is eating up a lot of JuliaLang's queue. Could you enable it on your fork and work things out on a different branch for now?

**Comment by tkelman** [comment]
We could also move this package to JuliaInterop.

**Comment by Keno** [comment]
Yep, will move to JuliaInterop

**Comment by ararslan** [comment]
I enabled AppVeyor here so if you have pyjulia under your AppVeyor account, you can modify it to point here. :)

File: .travis.yml [hunk 6]
```diff
@@ -46,6 +46,7 @@ matrix:
 notifications:
   email: false
 before_script:
+  - echo ./ci/install-julia.sh $JULIA_VERSION
```
Comment (tkelman) [inline .travis.yml]: if you end up needing to make changes to this script I'd still rather "upstream" them and have just one copy that lives in the website repo, once you get it all working

File: appveyor.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+install:
+# Download most recent Julia Windows binary
+  - ps: (new-object net.webclient).DownloadFile(
+        $("http://s3.amazonaws.com/"+$env:JULIAVERSION),
+        "C:\projects\julia-binary.exe")
+# Run installer silently, output to C:\projects\julia
+  - C:\projects\julia-binary.exe /S /D=C:\projects\julia
+
+build_script:
+  - "SET PATH=%PYTHONDIR%;%PYTHONDIR%\\Scripts;%PATH%"
```
Comment (tkelman) [inline appveyor.yml]: add the fast fail snippet for repeated pushes to the same pr here

Comment (Keno) [inline appveyor.yml]: What's the fast fail snippet?

Comment (tkelman) [inline appveyor.yml]: it's a powershell web request thing that should be in Example.jl's appveyor.yml

**Comment by Keno** [comment]
Windows test failure is due to it trying to find `rint` in the executable and of course not finding it since python is not linked against openlibm. I'm not sure why that's happening, but since it seems fixed on 0.5, I'm not going to put too much effort into attempting to figure out why.

File: fake-julia/julia.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+jl_base_module = c_void_p.in_dll(libjulia, "jl_base_module")
+_start = libjulia.jl_get_global(jl_base_module, libjulia.jl_symbol(b"_start"))
+args = (c_void_p * 1)()
+args[0] = _start
+libjulia.jl_apply_generic(args, 1)
+libjulia.jl_atexit_hook(0)
+
+# As an optimization, share precompiled packages with the main cache directory
+libjulia.jl_eval_string(b"""
+    outputji = Base.JLOptions().outputji
```
Comment (tkelman) [inline fake-julia/julia.py]: should this use a `let` to avoid always defining this?

Comment (Keno) [inline fake-julia/julia.py]: This is after the incremental image has been written, so it'll go away.

File: test/test_core.py [hunk 9]
```diff
@@ -44,11 +45,9 @@ def add(a, b):
             return a + b
         self.assertEqual([1, 4, 9],
                          list(julia.map(lambda x: x * x, [1, 2, 3])))
-        print(julia.map(lambda x: x + 1,
-                                   array.array('I', [10, 10, 10])))
-        self.assertEqual([11, 11, 11],
+        self.assertTrue(np.array_qual([11, 11, 11],
                          julia.map(lambda x: x + 1,
-                                   array.array('I', [10, 10, 10])))
+                                   array.array('I', [10, 10, 10]))))
```
Comment (ararslan) [inline test/test_core.py]: I can only imagine how Lisp programmers must feel all the time

**Comment by Keno** [comment]
Ok, I'm gonna go and merge this. Travis OS X is complaining, but that's an ecosystem wide problem right now, so oh well. If it turns out to be broken after, I'll fix that separately.

**Pull request merged**